### ISSUE TITLE
cluster: broadcast message to many actors of same kind

### ISSUE THREAD

**Issue by stuartcarnie** [AsynkronIT/protoactor-go#62]
Provide a new API for the cluster package to send an asynchronous message to a set of named, homogeneous actors.

`cluster.TellMany(names []string, kind string, message interface{})`

Two possible approaches to the implementation, with different pros / cons.

* C = client node
* P = partition node
* A = actor node

### Option 1

1. C: Hash all names, locating all partitions
1. C: request the location of all actors; wait for partition(s) to respond
    1. P: locate PIDs for all actors
    1. P: send PIDs to C
1. C: group by host
1. C: broadcast message to each host, A
    1. A: foreach local PID; forward message to target actor

**Pros**
* message for set of PIDs is delivered only once to each A node

**Cons**
* C must wait for P node to respond before it can forward to A nodes

### Option 2

1. C: Hash all names, locating all partitions
1. C: broadcast message to each partition P with subset of names
    1. P: foreach partition, locate PID for each name
    1. P: group by host and forward broadcast message to each node, A
        1. A: foreach local PID; forward message to target actor

**Pros** 
* broadcast is asynchronous, therefore node C does not have to wait

**Cons**
* message may be delivered multiple times to each A node for a subset of PIDs 
    * each partition P forwards on messages to each of its A nodes

### Option 3

Option 3 is a variation of Option 1, providing benefits of both Option 1 and 2

* B = client node broadcast actor, used for coordinating

1. C: send message to local "broadcast" actor B
    1. B: request the location of all actors
        1. P: locate PIDs for all actors
        1. P: replies with PIDs to B
    1. B: group by host
    1. B: broadcast message to each host, A
        1. A: foreach local PID; forward message to target actor

**Comment by stuartcarnie** [comment]
We could also expand the `TellMany` API to include an additional option controlling the delivery behavior if actor is not started:

* `ActorAutoStart`
* `ActorSkip`

**Comment by rogeralsing** [comment]
cc @ReubenBond,  @sergeybykov.
I assume you two have explored this area, batching/grouping broadcast messages for virtual actors across partitions and location nodes?

Thoughts? valuable? problems?

**Comment by sergeybykov** [comment]
We discussed similar ideas at length in https://github.com/dotnet/orleans/issues/610. I tried to spell it out in https://github.com/dotnet/orleans/issues/610#issuecomment-172921259. As was correctly pointed by @gabikliot, cases 1 and 2 semantically are pretty much equivalent to streams. Except we don't efficiently broadcast stream messages today.

So far we only implemented case 3 https://github.com/dotnet/orleans/issues/2444, which is not really about broadcasting, but was mentioned in that discussion.
